pkg/storage/chunk/cache/memcached.go

```diff
@@ -63,6 +63,9 @@ type Memcached struct {
 	stopped atomic.Bool
 
 	logger log.Logger
+
+	// NOTE: testFetchDelay should be used only for testing. See `SetTestFetchDelay()` method for more details.
+	testFetchDelay chan struct{}
```

Comment:
The idiomatic approach here would be to define a mock client or wrap this client and shadow the function where the delay is required.

If there's absolutely no other way to implement this without refactoring a lot then I'll approve, but this feels like a hack in its current form.

Reply:
> If there's absolutely no other way to implement this without refactoring a lot then I'll approve, but this feels like a hack in its current form.

Looks like it. I spent some time to clean it up. Tried several ways with mock as well. But there are some major blocks 

1. we use `cache_test` intead of `cache` (original package name) in all the tests to test these caches as external clients, which prevent us to use something internal for testing.

2. I tried to make it just `cache` touching tons of files, even after doing that lots of cyclic dependencies break this particular package. Then realized that's why they introduced this testing via `cache_test` in the first place.

3. Another core problem is in general with this package is how we do parallelism with Go routines. Ideally, only sender goroutine of the channel should be closing it as well (e.g: inputCh, resultsCh, etc). But here we are doing at random places. This prevents us to do some idiomatic tests deterministically. I tried to clean it up but ended up rewriting a lots of code, seems overkill for this PR. Better to do it separately with more testing.

Reply:
OK, are you planning to follow up this PR with the refactor?

Reply:
yes :)